src/components/ha-camera-stream.ts

```diff
@@ -87,54 +86,79 @@ export class HaCameraStream extends LitElement {
+      ? html`<ha-hls-player
+          autoplay
+          playsinline
+          .allowExoPlayer=${this.allowExoPlayer}
+          .muted=${this.muted}
+          .controls=${this.controls}
+          .hass=${this.hass}
+          .entityid=${this.stateObj.entity_id}
+          .posterUrl=${this._posterUrl}
+          @streams=${this._handleHlsStreams}
+          class=${!this._streamType && this._webRtcStreams ? "hidden" : ""}
+        ></ha-hls-player>`
+      : nothing}
+    ${this._streamType === STREAM_TYPE_WEB_RTC ||
+    (!this._streamType &&
+      this._capabilities?.frontend_stream_types.includes(STREAM_TYPE_WEB_RTC))
+      ? html`<ha-web-rtc-player
+          autoplay
+          playsinline
+          .muted=${this.muted}
+          .controls=${this.controls}
+          .hass=${this.hass}
+          .entityid=${this.stateObj.entity_id}
+          .posterUrl=${this._posterUrl}
+          @streams=${this._handleWebRtcStreams}
+          class=${this._streamType !== STREAM_TYPE_WEB_RTC &&
+          !this._webRtcStreams
+            ? "hidden"
+            : ""}
+        ></ha-web-rtc-player>`
+      : nothing}`;
+  }
+
+  private async _getCapabilities() {
+    this._capabilities = undefined;
+    this._hlsStreams = undefined;
+    this._webRtcStreams = undefined;
+    if (!supportsFeature(this.stateObj!, CAMERA_SUPPORT_STREAM)) {
+      return;
     }
-    if (this.stateObj.attributes.frontend_stream_type === STREAM_TYPE_WEB_RTC) {
-      return html`<ha-web-rtc-player
-        autoplay
-        playsinline
-        .muted=${this.muted}
-        .controls=${this.controls}
-        .hass=${this.hass}
-        .entityid=${this.stateObj.entity_id}
-        .posterUrl=${this._posterUrl}
-      ></ha-web-rtc-player>`;
+    this._capabilities = await fetchCameraCapabilities(
+      this.hass!,
+      this.stateObj!.entity_id
+    );
+    if (this._capabilities.frontend_stream_types.length === 1) {
+      this._streamType = this._capabilities.frontend_stream_types[0];
     }
-    return nothing;
   }
 
   private get _shouldRenderMJPEG() {
-    if (this._forceMJPEG === this.stateObj!.entity_id) {
-      // Fallback when unable to fetch stream url
-      return true;
-    }
     if (!supportsFeature(this.stateObj!, CAMERA_SUPPORT_STREAM)) {
       // Steaming is not supported by the camera so fallback to MJPEG stream
       return true;
     }
     if (
-      this.stateObj!.attributes.frontend_stream_type === STREAM_TYPE_WEB_RTC
+      this._capabilities &&
+      (!this._capabilities.frontend_stream_types.includes(STREAM_TYPE_HLS) ||
+        this._hlsStreams?.hasVideo === false) &&
+      (!this._capabilities.frontend_stream_types.includes(
+        STREAM_TYPE_WEB_RTC
+      ) ||
+        this._webRtcStreams?.hasVideo === false)
     ) {
-      // Browser support required for WebRTC
-      return typeof RTCPeerConnection === "undefined";
+      // No video in HLS stream and no video in WebRTC stream
+      return true;
```

Comment:
@bramkragten I think this change might have caused a regression in the HLS player. When the `_setRetryableError` is called from the `ha-hls-player` with:

```js
  private _setRetryableError(errorMessage: string) {
    this._error = errorMessage;
    this._errorIsFatal = false;
    fireEvent(this, "streams", { hasAudio: false, hasVideo: false });
  }
```

This code path (or the updated path from https://github.com/home-assistant/frontend/pull/22674) will cause the camera stream to switch to the `mjpeg` stream without retrying. I have a situation where an intermittent `bufferAppendError` being raised from an HLS stream, and instead of retrying it immediately falls back to the mjpeg stream. Previously I think that fallback would only have happened if `fetchStreamUrl` failed.

What do you think?